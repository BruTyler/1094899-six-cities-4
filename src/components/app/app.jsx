import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {BrowserRouter, Route, Switch} from 'react-router-dom';
import {connect} from 'react-redux';

import {ActionCreator} from '../../reducer/application/application.js';
import {getCity, getApartmentList, getCities} from '../../reducer/data/selectors.js';
import MainScreen from './../main-screen/main-screen.jsx';
import Property from '../property/property.jsx';
import withActiveItem from '../../hocs/with-active-item/with-active-item.jsx';
import OfflineScreen from '../offline-screen/offline-screen.jsx';

class App extends PureComponent {
  constructor(props) {
    super(props);
    this._init();
  }

  _init() {
    const {handleFirstCityLoad, cityList} = this.props;
    if (cityList && cityList.length > 0) {
      handleFirstCityLoad(cityList[0]);
    }
  }

  _renderScreen() {
    const {apartmentList, cityList, activeCity, onCityTitleClick,
      activeItem: clickedProperty,
      onItemSelect: onApartmentTitleClick
    } = this.props;

    if (activeCity === undefined || cityList === undefined || cityList.length === 0) {
      return <OfflineScreen />;
    }

    if (clickedProperty) {
      return <Property
        apartment={clickedProperty}
        neighboorApartmentList={apartmentList.filter((el) => el.id !== clickedProperty.id).slice(0, 3)}
        city={activeCity}
        onApartmentTitleClick={onApartmentTitleClick}
      />;
    }

    return <MainScreen
      activeCity={activeCity}
      cityList={cityList}
      apartmentList={apartmentList}
      onApartmentTitleClick={onApartmentTitleClick}
      onCityTitleClick={onCityTitleClick}
    />;
  }

  render() {
    return <BrowserRouter>
      <Switch>
        <Route exact path="/">
          {this._renderScreen()}
        </Route>
      </Switch>
      <Switch>
        <Route exact path="/offline">
          <OfflineScreen />
        </Route>
      </Switch>
    </BrowserRouter>;
  }
}

App.propTypes = {
  apartmentList: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  cityList: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  activeCity: PropTypes.shape(),
  onCityTitleClick: PropTypes.func.isRequired,
  activeItem: PropTypes.shape(),
  onItemSelect: PropTypes.func.isRequired,
  handleFirstCityLoad: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    apartmentList: getApartmentList(state),
    cityList: getCities(state),
    activeCity: getCity(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  onCityTitleClick(city) {
    dispatch(ActionCreator.changeCity(city.id));
  },
  handleFirstCityLoad(city) {
    dispatch(ActionCreator.changeCity(city.id));
  },
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(withActiveItem(App));
