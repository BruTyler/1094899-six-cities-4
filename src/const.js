export const PlaceType = {
  APARTMENT: `Apartment`,
  PRIVATE_ROOM: `Private room`,
};
