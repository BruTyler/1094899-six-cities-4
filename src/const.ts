export enum PlaceType {
  APARTMENT = `Apartment`,
  ROOM = `Private room`,
  HOUSE = `House`,
  HOTEL = `Hotel`,
};

export enum SortType {
  POPULAR = `Popular`,
  PRICE_LOW = `Price: low to high`,
  PRICE_HIGH = `Price: high to low`,
  TOP_RATED = `Top rated first`,
};

export enum AuthorizationStatus {
  NO_AUTH = `NO_AUTH`,
  AUTH = `AUTH`,
};

export enum BuisnessRequirements {
  MAX_PHOTOS_PER_APARTMENT = 6,
  MAX_REVIEWS_PER_APARTMENT = 10,
  MIN_REVIEW_TEXT_LENGTH = 50,
  MAX_REVIEW_TEXT_LENGTH = 300,
};

export enum AppRoute {
  AUTH = `/login`,
  ROOT = `/`,
  FAVORITES = `/favorites`,
  PROPERTY_WITH_ID = `/offer/:id`,
};

export enum ApartmentEnvironment {
  FAVORITE = `favorites`,
  NEARBY_PLACES = `near-places`,
  MAIN_WINDOW = `cities`
};

export enum MapEnvironment {
  NEARBY_PLACES = `property__map`,
  MAIN_WINDOW = `cities__map`
};
