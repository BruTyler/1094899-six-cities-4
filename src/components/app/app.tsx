import * as React from 'react';
import {Route, Switch, Router} from 'react-router-dom';
import {connect} from 'react-redux';

import {ActionCreator} from '../../reducer/application/application';
import {getCity, getApartmentList, getCities} from '../../reducer/data/selectors';
import MainScreen from './../main-screen/main-screen';
import Property from '../property/property';
import {Operation as UserOperation} from '../../reducer/user/user';
import AuthScreen from '../auth-screen/auth-screen';
import {getAuthInfo} from '../../reducer/user/selectors';
import {AppRoute} from '../../const';
import history from '../../history';
import FavoriteScreen from '../favorite-screen/favorite-screen';
import PrivateRoute from '../private-route/private-route';
import {City, AuthInfo, AuthData, SingleApartment} from '../../types';

interface Props {
  apartmentList: Array<SingleApartment>;
  cityList: Array<City>;
  activeCity: City;
  handleChangeCity: (city: City) => void;
  handleLoginSubmit: (authData: AuthData) => Promise<void>;
  authInfo?: AuthInfo;
}

const App: React.FunctionComponent<Props> = (props: Props) => {
  const {authInfo, handleLoginSubmit,
    apartmentList, cityList, activeCity, handleChangeCity
  } = props;

  return <Router history={history}>
    <Switch>
      <Route exact path={AppRoute.AUTH}>
        <AuthScreen
          authInfo={authInfo}
          onLoginSubmit={handleLoginSubmit}
          activeCity={activeCity}
        />
      </Route>
      <PrivateRoute exact
        path={AppRoute.FAVORITES}
        render={() => <FavoriteScreen />}
      />
      <Route exact
        path={AppRoute.PROPERTY_WITH_ID}
        render={({match}) =>
          <Property
            id={Number(match.params.id)}
          />}
      />
      <Route path={AppRoute.ROOT}>
        <MainScreen
          activeCity={activeCity}
          cityList={cityList}
          apartmentList={apartmentList}
          onCityTitleClick={handleChangeCity}
          authInfo={authInfo}
        />;
      </Route>
    </Switch>
  </Router>;
};

const mapStateToProps = (state) => {
  return {
    apartmentList: getApartmentList(state),
    cityList: getCities(state),
    activeCity: getCity(state),
    authInfo: getAuthInfo(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  handleLoginSubmit(authData) {
    return dispatch(UserOperation.makeAuthorization(authData));
  },
  handleChangeCity(city) {
    dispatch(ActionCreator.changeCity(city.id));
  },
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
