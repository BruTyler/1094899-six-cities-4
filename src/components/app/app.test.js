import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import {App} from './app.jsx';
import {SortType, AuthorizationStatus} from '../../const.js';
import NameSpace from '../../reducer/name-space.js';

const EMPTY_HANDLER = () => {};
const CITIES = [
  {
    id: `A`,
    location: [1, 2],
    defaultZoom: 1
  },
  {
    id: `B`,
    location: [3, 4],
    defaultZoom: 2
  },
];

const APARTMENTS = [
  {
    id: 0,
    type: `Apartment`,
    description: `description0`,
    rating: 0.1,
    price: 1,
    isPremium: true,
    isFavourite: true,
    photo: `img0.jpg`,
    photoSet: [`img2.jpg`],
    bedrooms: 3,
    adultsMax: 4,
    goods: [`WiFi`, `TV`],
    host: {
      avatar: `ava.jpg`,
      name: `name`,
      isSuper: true,
    },
    location: [1, 2],
  },
];
const mockStore = configureStore();

describe(`<App /> render suit`, () => {
  it(`<App /> render apartment list`, () => {
    const store = mockStore({
      [NameSpace.DATA]: {
        apartmentList: APARTMENTS,
      },
      [NameSpace.APPLICATION]: {
        sortType: SortType.POPULAR,
      },
    });

    const generatedTree = renderer.create(
        <Provider store={store}>
          <App
            cityList={CITIES}
            activeCity={CITIES[0]}
            apartmentList={APARTMENTS}
            handleChangeCity={EMPTY_HANDLER}
            handleFirstCityLoad={EMPTY_HANDLER}
            handleFetchingHotels={EMPTY_HANDLER}
            handleFinishLoading={EMPTY_HANDLER}
            onLoginSubmit={EMPTY_HANDLER}
            authStatus={AuthorizationStatus.NO_AUTH}
          />
        </Provider>,
        {
          createNodeMock: () => {
            return document.createElement(`div`);
          }
        }
    ).toJSON();

    expect(generatedTree).toMatchSnapshot();
  });
});
