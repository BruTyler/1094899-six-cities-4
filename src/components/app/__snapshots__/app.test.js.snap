// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render suit App render apartment list 1`] = `
Array [
  <p>
    Предложения аренды:
  </p>,
  <div>
    <img
      height="40"
      src="img/no-places.png"
      width="50"
    />
    <a
      className="apartment__title"
      href="#"
      onClick={[Function]}
    >
      apart1
    </a>
  </div>,
  <div>
    <img
      height="40"
      src="img/no-places.png"
      width="50"
    />
    <a
      className="apartment__title"
      href="#"
      onClick={[Function]}
    >
      apart2
    </a>
  </div>,
  <hr />,
  <p>
    Всего предложений: 
    2
  </p>,
]
`;

exports[`App render suit App render empty apartment list 1`] = `
Array [
  <p>
    Предложения аренды:
  </p>,
  <hr />,
  <p>
    Всего предложений: 
    0
  </p>,
]
`;
